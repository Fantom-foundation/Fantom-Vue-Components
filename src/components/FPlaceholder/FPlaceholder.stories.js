import { withA11y } from '@storybook/addon-a11y';

import FPlaceholder from './FPlaceholder.vue';

export default {
    title: 'FPlaceholder',
    component: FPlaceholder,
    decorators: [withA11y],
};

const testContStyle = {
    maxWidth: '350px',
    // border: '1px solid #eee',
    // padding: '16px',
    margin: '16px',
};

const testContStyle2 = {
    maxWidth: '850px',
    margin: '16px',
};

const TIMEOUT = 2000;

export const Default = () => ({
    components: { FPlaceholder },
    template: `
        <div>
            <div :style="testContStyle">
                Lorem ipsum dolor sit amet,
                <b><f-placeholder :content-loaded="!!item"><em>{{ item }}</em></f-placeholder></b>
                consectetur adipisicing elit.
            </div>
            <button class="btn light" @click="init">Go</button>
        </div>
    `,
    created() {
        this.init();
    },
    data() {
        return {
            item: '',
            testContStyle,
        };
    },
    methods: {
        init() {
            this.item = '';

            setTimeout(() => {
                this.item = 'foo';
            }, TIMEOUT);
        },
    },
});

export const ReplacementText = () => ({
    components: { FPlaceholder },
    template: `
        <div>
            <div :style="testContStyle">
                Lorem ipsum dolor sit amet,
                <b><f-placeholder replacement-text="123" :content-loaded="!!item"><em>{{ item }}</em></f-placeholder></b>
                consectetur adipisicing elit.
            </div>
            <button class="btn light" @click="init">Go</button>
        </div>
    `,
    created() {
        this.init();
    },
    data() {
        return {
            item: '',
            testContStyle,
        };
    },
    methods: {
        init() {
            this.item = '';

            setTimeout(() => {
                this.item = 'foo';
            }, TIMEOUT);
        },
    },
});

export const LongerReplacementText = () => ({
    components: { FPlaceholder },
    template: `
        <div>
            <div :style="testContStyle">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder
                        replacement-text="Lorem ipsum dolor sit amet, consectetur adipisicing elit."
                        :content-loaded="!!item"
                    >
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <button class="btn light" @click="init">Go</button>
        </div>
    `,
    created() {
        this.init();
    },
    data() {
        return {
            item: '',
            testContStyle,
        };
    },
    methods: {
        init() {
            this.item = '';

            setTimeout(() => {
                this.item = 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.';
            }, TIMEOUT);
        },
    },
});

export const ReplacementNumChars = () => ({
    components: { FPlaceholder },
    template: `
        <div>
            <div :style="testContStyle">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder :replacement-num-chars="71" :content-loaded="!!item">
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <button class="btn light" @click="init">Go</button>
        </div>
    `,
    created() {
        this.init();
    },
    data() {
        return {
            item: '',
            testContStyle,
        };
    },
    methods: {
        init() {
            this.item = '';

            setTimeout(() => {
                this.item = 'Aspernatur assumenda aut culpa, cupiditate dolor eius enim facere fuga.';
            }, TIMEOUT);
        },
    },
});

export const Block = () => ({
    components: { FPlaceholder },
    template: `
        <div>
            <div :style="testContStyle">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder block :replacement-num-chars="120" :content-loaded="!!item">
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <div :style="testContStyle2">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder block :replacement-num-chars="140" :content-loaded="!!item">
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <button class="btn light" @click="init">Go</button>
        </div>
    `,
    created() {
        this.init();
    },
    data() {
        return {
            item: '',
            testContStyle,
            testContStyle2,
        };
    },
    methods: {
        init() {
            this.item = '';

            setTimeout(() => {
                this.item =
                    'Aspernatur assumenda aut culpa, cupiditate dolor eius enim facere fuga. Lorem ipsum dolor sit amet, consectetur adipisicing elit.';
            }, TIMEOUT);
        },
    },
});

export const Animation = () => ({
    components: { FPlaceholder },
    template: `
        <div>
            <div :style="testContStyle">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder animation="fplaceholder-pulsebganim" :replacement-num-chars="120" block :content-loaded="!!item">
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <div :style="testContStyle2">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder animation="fplaceholder-pulsebganim" :replacement-num-chars="140" block :content-loaded="!!item">
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <button class="btn light" @click="init">Go</button>
        </div>
    `,
    created() {
        this.init();
    },
    data() {
        return {
            item: '',
            testContStyle,
            testContStyle2,
        };
    },
    methods: {
        init() {
            this.item = '';

            setTimeout(() => {
                this.item =
                    'Aspernatur assumenda aut culpa, cupiditate dolor eius enim facere fuga. Lorem ipsum dolor sit amet, consectetur adipisicing elit.';
            }, TIMEOUT);
        },
    },
});

export const NoAnimation = () => ({
    components: { FPlaceholder },
    template: `
        <div>
            <div :style="testContStyle">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder :use-animation="false" :replacement-num-chars="120" block :content-loaded="!!item">
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <div :style="testContStyle2">
                Lorem ipsum dolor sit amet,
                <b>
                    <f-placeholder :use-animation="false" :replacement-num-chars="140" block :content-loaded="!!item">
                        <em>{{ item }}</em>
                    </f-placeholder>
                </b>
                consectetur adipisicing elit.
            </div>
            <button class="btn light" @click="init">Go</button>
        </div>
    `,
    created() {
        this.init();
    },
    data() {
        return {
            item: '',
            testContStyle,
            testContStyle2,
        };
    },
    methods: {
        init() {
            this.item = '';

            setTimeout(() => {
                this.item =
                    'Aspernatur assumenda aut culpa, cupiditate dolor eius enim facere fuga. Lorem ipsum dolor sit amet, consectetur adipisicing elit.';
            }, TIMEOUT);
        },
    },
});
