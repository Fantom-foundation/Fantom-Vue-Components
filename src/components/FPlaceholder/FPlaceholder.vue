<template>
    <span class="fplaceholder" :class="classes">
        <template v-if="contentLoaded"><slot></slot></template>
        <template v-else>
            <span v-if="!block" ref="twin" class="fplaceholder_replacementtwin">{{ replacement || '&nbsp;' }}</span>
            <span class="fplaceholder_replacement">{{ replacement || '&nbsp;' }}</span>
        </template>
    </span>
</template>

<script>
/**
 * Placeholder for a content. Waits for `contentLoaded` to be `true`, meanwhile it shows not visible replacement element
 * with background animation.
 */
export default {
    name: 'FPlaceholder',

    props: {
        /** Determines whether the content is loaded or not */
        contentLoaded: {
            type: Boolean,
            default: false,
        },
        /** Text to be used in replacement element (not visible) */
        replacementText: {
            type: String,
            default: '',
        },
        /** Number of characters of placeholder text "Lorem ipsum ..." used as a replacement text, if `replacementText` is not specified */
        replacementNumChars: {
            type: Number,
            default: 0,
            validator: function(value) {
                return value >= 0;
            },
        },
        animation: {
            type: String,
            default: 'fplaceholder-movingbganim',
        },
        /** Set `animation` class while content is loading */
        useAnimation: {
            type: Boolean,
            default: true,
        },
        /** Display placeholder as a block */
        block: {
            type: Boolean,
            default: false,
        },
    },

    data() {
        return {
            loremIpsum:
                'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aliquid ipsam magni nam odit praesentium temporibus vel voluptates voluptatibus. Aliquam atque autem cum dolores iure minus officia quibusdam quis vel voluptates!</div><div>Dicta numquam officiis placeat sed tempore? Consequatur corporis deserunt doloribus, eaque earum enim, ex facere fugiat, hic id iusto minima nisi obcaecati perferendis repellat sunt tempora voluptate! Distinctio, laborum quia.</div><div>Architecto debitis dicta doloremque dolores eligendi enim error exercitationem, explicabo fugiat hic labore laudantium molestias nesciunt nihil obcaecati porro voluptas voluptate voluptatem. Blanditiis, culpa dignissimos odio porro provident quam velit!',
        };
    },

    computed: {
        replacement() {
            let replacement = this.replacementText;

            if (replacement === '' && this.replacementNumChars > 0) {
                replacement = this.loremIpsum.slice(0, this.replacementNumChars);
            }

            return replacement;
        },
        classes() {
            return {
                'fplaceholder-on': !this.contentLoaded,
                'fplaceholder-isblock': this.block,
                [this.animation]: this.useAnimation && !this.contentLoaded,
            };
        },
    },

    mounted() {
        this.$nextTick(() => {
            this.setBackgroundSize();
        });
    },

    methods: {
        setBackgroundSize() {
            if (this.animation !== 'fplaceholder-movingbganim') {
                return;
            }

            let width = 0;

            if (this.block) {
                width = this.$el.clientWidth;
            } else if (!this.contentLoaded) {
                const eTwin = this.$refs.twin;

                if (eTwin) {
                    const elStyle = window.getComputedStyle(this.$el);

                    eTwin.style.padding = elStyle.padding;
                    eTwin.style.fontSize = elStyle.fontSize;
                    eTwin.style.fontFamily = elStyle.fontFamily;

                    width = eTwin.clientWidth;
                }
            }

            if (width > 0) {
                this.$el.style.setProperty('--fplaceholder-background-size-width', `${width * 2}px`);
            }
        },
    },
};
</script>

<style lang="scss">
@use 'style';
</style>
