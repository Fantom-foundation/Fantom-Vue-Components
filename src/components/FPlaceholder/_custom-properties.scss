.fplaceholder {
    --fplaceholder-background-size-width: 800px;
    --fplaceholder-bg-color1: transparent;
    --fplaceholder-bg-color2: #f4f4f4;
    --fplaceholder-animation-speed: 1s;
}
