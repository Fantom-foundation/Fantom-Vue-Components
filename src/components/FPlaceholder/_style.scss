@use 'custom-properties';

@keyframes moving-background-animation {
    0% {
        background-position: 100% 0;
    }
    100% {
        background-position: -100% 0;
    }
}

@keyframes pulse-background-animation {
    50% {
        background: transparent;
    }
}

.fplaceholder {
    &-on {
        .fplaceholder_replacement {
            opacity: 0;
        }
    }

    &-movingbganim {
        background: linear-gradient(
            to right,
            var(--fplaceholder-bg-color1) 8%,
            var(--fplaceholder-bg-color2) 18%,
            var(--fplaceholder-bg-color1) 33%
        );
        background-size: var(--fplaceholder-background-size-width) 100%;

        animation-duration: var(--fplaceholder-animation-speed);
        animation-fill-mode: forwards;
        animation-iteration-count: infinite;
        animation-name: moving-background-animation;
        animation-timing-function: linear;
    }

    &-pulsebganim {
        background-color: var(--fplaceholder-bg-color2);

        animation-duration: var(--fplaceholder-animation-speed);
        animation-iteration-count: infinite;
        animation-name: pulse-background-animation;
        animation-timing-function: linear;
    }

    &-isblock {
        display: block;

        .fplaceholder_replacement {
            display: block;
        }
    }

    &_replacementtwin {
        position: fixed;
        top: -1000px;
        left: -1000px;
        visibility: hidden;
        white-space: pre;
    }
}
