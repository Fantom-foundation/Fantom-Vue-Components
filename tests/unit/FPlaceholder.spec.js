import { mount } from '@vue/test-utils';
import FPlaceholder from '@/components/FPlaceholder/FPlaceholder.vue';

const REPLACEMENT_TEXT = 'Replacement text';
const DEFAULT_SLOT_TEXT = '123456';
let wrapper = null;

function destroyWrapper(wrapper) {
    if (wrapper) {
        wrapper.destroy();
        wrapper = null;
    }
}

function createWrapper({ propsData = {}, slots = { default: DEFAULT_SLOT_TEXT } } = {}) {
    return mount(FPlaceholder, {
        propsData,
        slots,
        // attachTo: document.body,
    });
}

afterEach(() => {
    destroyWrapper(wrapper);
});

describe('FPlaceholder', () => {
    it('should render correctly', () => {
        wrapper = createWrapper({ slots: {} });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('should set `replacementText` as a replacement content if `contentLoaded` is `false` and then content from default slot if `contentLoaded` is `true`', async () => {
        wrapper = createWrapper({ propsData: { replacementText: REPLACEMENT_TEXT } });

        expect(wrapper.find('.fplaceholder_replacementtwin').text()).toBe(REPLACEMENT_TEXT);

        await wrapper.setProps({ contentLoaded: true });

        expect(wrapper.text()).toBe(DEFAULT_SLOT_TEXT);
    });

    it('should extract a `replacementNumChars` number of characters of "Lorem ipsum ..." as a replacement text', () => {
        wrapper = createWrapper({ propsData: { replacementNumChars: 11 } });

        expect(wrapper.find('.fplaceholder_replacementtwin').text()).toBe('Lorem ipsum');
    });

    it('should set given animation class if content is loading and `animation` prop is specified', () => {
        wrapper = createWrapper({ propsData: { animation: 'anim' } });

        expect(wrapper.classes()).toContain('anim');
    });

    it('should not set given animation class if content is loading and `useAnimation` is set to `false`', () => {
        wrapper = createWrapper({ propsData: { useAnimation: false } });

        expect(wrapper.classes()).not.toContain(wrapper.props('animation'));
    });
});
