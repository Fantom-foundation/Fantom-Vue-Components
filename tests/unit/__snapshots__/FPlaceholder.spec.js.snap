// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FPlaceholder should render correctly 1`] = `
<span
  class="fplaceholder fplaceholder-on fplaceholder-movingbganim"
>
  <span
    class="fplaceholder_replacementtwin"
  >
     
  </span>
   
  <span
    class="fplaceholder_replacement"
  >
     
  </span>
</span>
`;
